.github/workflows/pr-update-playwright-expectations.yaml

```diff
@@ -12,74 +12,167 @@ concurrency:
+    needs: setup
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        shardIndex: [1, 2, 3, 4]
+        shardTotal: [4]
+    steps:
+      # Restore test environment from setup job
+      - name: Wait for cache propagation
+        run: sleep 10
+
+      - name: Restore cached setup
+        uses: actions/cache/restore@5a3ec84eff668545956fd18022155c47e93e2684
+        with:
+          fail-on-cache-miss: true
+          path: .
+          key: comfyui-setup-${{ needs.setup.outputs.cache-key }}
+
+      - name: Setup ComfyUI server
         uses: ./.github/actions/setup-comfyui-server
         with:
           launch_server: true
+
+      - name: Setup nodejs, pnpm, reuse built frontend
+        uses: ./.github/actions/setup-frontend
+
       - name: Setup Playwright
         uses: ./.github/actions/setup-playwright
-      - name: Run Playwright tests and update snapshots
+
+      # Run sharded tests with snapshot updates
+      - name: Update snapshots (Shard ${{ matrix.shardIndex }}/${{ matrix.shardTotal }})
         id: playwright-tests
-        run: pnpm exec playwright test --update-snapshots
+        run: pnpm exec playwright test --update-snapshots --shard=${{ matrix.shardIndex }}/${{ matrix.shardTotal }}
         continue-on-error: true
-      - uses: actions/upload-artifact@v4
+
+      # Upload updated snapshots from this shard
+      - name: Upload updated snapshots
+        uses: actions/upload-artifact@v4
+        if: always()
+        with:
+          name: snapshots-shard-${{ matrix.shardIndex }}
+          path: browser_tests/**/*-snapshots/
+          retention-days: 1
+
+      - name: Upload test report
+        uses: actions/upload-artifact@v4
         if: always()
         with:
-          name: playwright-report
+          name: playwright-report-shard-${{ matrix.shardIndex }}
           path: ./playwright-report/
           retention-days: 30
+
+  # Merge snapshots and commit
+  merge-and-commit:
+    needs: [setup, update-snapshots-sharded]
+    runs-on: ubuntu-latest
+    if: always()
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v5
+        with:
+          ref: ${{ needs.setup.outputs.branch }}
+
+      # Download all snapshot artifacts from shards
+      - name: Download all snapshots
+        uses: actions/download-artifact@v4
+        with:
+          pattern: snapshots-shard-*
+          path: ./downloaded-snapshots
+          merge-multiple: false
+
+      # Merge snapshots from all shards into browser_tests directory
+      - name: Merge snapshots
+        run: |
+          echo "Merging snapshots from all shards..."
+          for shard_dir in ./downloaded-snapshots/snapshots-shard-*; do
+            if [ -d "$shard_dir" ]; then
+              echo "Copying from $shard_dir..."
+              cp -r "$shard_dir"/* browser_tests/ 2>/dev/null || true
+            fi
+          done
+          echo "Snapshot merge complete"
```

Comment:
For some reason, this part worries me.
Can we fail loudly here instead of sending stderr and failed statuses to the cornfield?